src/lib/loaders/batchLoader.ts

```diff
@@ -20,35 +62,35 @@ const renderParams = key => {
     .join("&")
 }
 
-interface GroupKeysResult {
-  id: string[]
-  size: number
-  [key: string]: any
-}
-export const groupKeys = (
-  requestedKeys: string | { id }
-): GroupKeysResult[] => {
-  const grouped = chain(requestedKeys)
-    .groupBy(renderParams)
-    .values()
-    .map(values => chunk(values, 20))
-    .flatten()
-    .map((keys: string[] | { id }[]) => {
-      if (typeof keys[0] === "string") {
-        return { id: keys, size: keys.length }
-      }
-      return {
-        // @ts-ignore
-        ...keys[0],
-        id: keys.map(k => k.id),
-        size: keys.length,
+/**
+ * Takes an array of normalized keys and groups them based on their parameters.
+ * @returns a tuple of an array of group strings and an array of grouped keys
+ */
+export const groupKeys = (requestedKeys): [KeyGroupList, NormalKey[][]] => {
+  const [keyGroupList, groupedKeys] = chain(requestedKeys)
+    .groupBy(getKeyGroup)
+    .entries()
+    .thru(entries => {
+      let result: [KeyGroupList, NormalKey[][]] = [[], []]
+      for (let i = 0; i < entries.length; ++i) {
+        result[0].push(entries[i][0])
+        result[1].push(entries[i][1])
       }
+      return result
     })
     .value()
 
-  return (grouped as unknown) as GroupKeysResult[]
+  return [keyGroupList, groupedKeys]
 }
 
+/**
+ * Collects all the params into one object to be passed to gravity loader
+ */
+export const compactKeyForRequest = (keys: NormalKey[]): CompactedKey => ({
+  ...keys[0],
+  id: keys.map(key => key.id),
+})
```

Comment:
I have a hard time understanding the goal of this function. The singular `id` key holds an array of ids, but otherwise  only the first key is used for the rest of the compacted key's traits? Why is that done in that way?

Reply:
The "key" is ultimately parameters that are being passed to gravity loader. To pass multiple ids to a list endpoint you turn id into an array of keys. The goal is to turn a list of like parameters (i.e. they all have the same parameters besides id) into a single object with a list of ids that the list endpoint can understand.

Reply:
Ah yes, ofc 🤦‍♂️ Thanks!